discourse/discourse
# FIX: When moving posts between threads, ensure only relevant timings are moved

## ✨ What's This?

When moving posts to an existing topic whilst preserving chronological ordering, the second chunk of posts moved can trigger a PostgreSQL error:

```
PG::UniqueViolation (ERROR:  duplicate key value violates unique constraint "post_timings_unique"
DETAIL:  Key (topic_id, post_number, user_id)=(36, 29, 1) already exists.
)
lib/mini_sql_multisite_connection.rb:109:in `run'
app/models/post_mover.rb:432:in `copy_shifted_post_timings_from_temp'
app/models/post_mover.rb:118:in `handle_moved_references'
app/models/post_mover.rb:94:in `move_posts_to'
app/models/post_mover.rb:28:in `block in to_topic'
app/models/post_mover.rb:28:in `to_topic'
```

The cause of this error is a little bit silly.

After a post has been moved from a different topic, a moderator post is inserted into the old topic, pointing to where the post has been moved to.

This moderator post is ultimately created using `PostCreator`, which by default, records a 5000ms Post Timing for that newly created post. Because of how Post Timings are tracked, this timing will be inserted into the `post_timings` table with the same `topic_id`, `post_number`, and `user_id` as the moved post used to have before it was moved.

Now, when we proceed to moving the next post, `PostMover.move_post_timings` will look for post timings to move based on whether there's a corresponding entry in the `moved_posts` table. This is where our problem lays: the post that we previously moved has an entry in the `moved_posts` table, but now there's an unexpected entry in `post_timings` that we know doesn't belong to the previously moved post, but it has matching `topic_id`, `post_number`, and `user_id` of the what that moved post used to have, anyway.

To fix this, we can enforce only matching post IDs that are being moved in this run of `PostMover`.

Internal ref t/142010/5

## 👑 Testing

The simplest way to reproduce this error is:

- Create a destination topic.
- Create the first source topic, and add a reply to it.
- Create a second source topic, and add a reply to it.
- Move the reply in the second source topic to the destination topic, remembering to check the “preserve chronological order” option.
- Move the reply in the first source topic to the destination, again checking the “preserve chronological order” option.
